source/bin/rocprof-sys-instrument/details.cpp

```diff
@@ -488,6 +488,77 @@ find_function(image_t* app_image, const std::string& _name, const strset_t& _ext
     return _func;
 }
 
+//======================================================================================//
+//
+//  Find undefined function symbols (external references) in the binary
+//
+symtab_symbol_t*
+find_undefined_function_symbol(image_t* app_image, const std::string& _name)
+{
+    if(_name.empty()) return nullptr;
+
+    // Get all objects from the image
+    BPatch_Vector<BPatch_object*> app_objects;
+    app_image->getObjects(app_objects);
+
+    if(app_objects.empty())
+    {
+        verbprintf(3, "No objects found in image for symbol search\n");
+        return nullptr;
+    }
+    // Search helper lambda for code reuse
+    auto _find_symbol = [](SymTab::Symtab*    symtab,
+                           const std::string& target_name) -> symtab_symbol_t* {
+        if(!symtab) return nullptr;
+
+        std::vector<SymTab::Symbol*> all_symbols;
+        if(!symtab->getAllSymbols(all_symbols)) return nullptr;
+
+        for(auto* symbol : all_symbols)
+        {
+            if(!symbol || symbol->getType() != SymTab::Symbol::ST_FUNCTION ||
+               symbol->getRegion())
+                continue;
+
+            // Try all possible symbol name representations
+            std::string symbol_name = symbol->getPrettyName();
+            if(symbol_name.empty()) symbol_name = symbol->getMangledName();
+            if(symbol_name.empty()) symbol_name = symbol->getTypedName();
+
+            // Check for exact match and undefined function criteria
+            if(symbol_name == target_name) return symbol;
+        }
+        return nullptr;
```

Comment:
```suggestion 
    auto symbol_iter = std::find_if(
        all_symbols.begin(), all_symbols.end(),
        [&target_name](SymTab::Symbol* symbol) {
            if (!symbol || symbol->getType() != SymTab::Symbol::ST_FUNCTION || symbol->getRegion())
                return false;

            std::string symbol_name = symbol->getPrettyName();
            if (symbol_name.empty()) symbol_name = symbol->getMangledName();
            if (symbol_name.empty()) symbol_name = symbol->getTypedName();

            return symbol_name == target_name;
        }
    );

    return symbol_iter != all_symbols.end() ? *symbol_iter : nullptr;
```

Maybe by doing this, we can have a clearer intention

Reply:
And you can apply the same approach for `for` loop below. With `find_if` we can also get rid of return statement in for loop.